packages/ragbits-conversations/src/ragbits/conversations/piepline/pipeline.py

```diff
@@ -0,0 +1,70 @@
+from collections.abc import Sequence
+
+from ragbits.conversations.piepline.state import ConversationPipelineResult, ConversationPipelineState
+from ragbits.conversations.piepline.state_to_chat import DefaultStateToChatStep
+from ragbits.conversations.piepline.steps import StrToStateStep
+from ragbits.core.llms.base import LLM
+from ragbits.core.pipelines.pipeline import Pipeline, Step
+from ragbits.core.prompt.prompt import ChatFormat
+
+
+class RunLLMStep(Step):
+    """
+    A plugin that runs the LLM on the conversation pipeline state.
+    """
+
+    def __init__(self, llm: LLM, state_to_chat: Step[ConversationPipelineState, ChatFormat]) -> None:
+        self.llm = llm
+        self.state_to_chat = state_to_chat
+
+    async def run(self, state: ConversationPipelineState) -> ConversationPipelineResult:
+        """
+        Processes the conversation pipeline state and returns the updated result.
+        """
+        chat = await self.state_to_chat.run(state)
+        stream = self.llm.generate_streaming(chat)
+        return ConversationPipelineResult(plugin_metadata=state.plugin_metadata, output_stream=stream)
+
+
+class ConversationPiepline(Step):
+    """
+    Class that runs a conversation pipeline with the given plugins
+    """
+
+    def __init__(
+        self,
+        llm: LLM,
+        preprocessors: Sequence[Step[ConversationPipelineState, ConversationPipelineState]] = [],
```

Comment:
Note: in this implementation you don't have plugins that can simulanously act on different stages of the pipeline (like you do with implementation 1). Here, if supporting a functionality requires both pre- and post- processing, you will need to have two different implementations and pass both as objects.